src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CustomFx2NLParams.cs

```diff
@@ -20,7 +20,12 @@ public class CustomFx2NLParams : BaseNLParams, IHasTextDocument
         /// <summary>
         /// Existing Power Fx Expression.
         /// </summary>
-        public string Expression { get; set; }
+        public string Expression { get; set; }
+
+        /// <summary>
+        /// Optional range of the expression.
+        /// </summary>
+        public Range Range { get; set; } = null;
```

Comment:
Are you sure about using Ranges? PowerFx uses one dimensional indexing. It would be expensive to convert 2d ranges to 1d indices on the server side as you would need to go through whole expression. If you make this 1d, (start and end index both are 1d) then you can convert the 2d indices on the client side very cheaply with a better time complexity. This is not a native protocol so there's nothing that binds us with 2d ranges. Consider passing flatten 1d index into the expression to avoid any conversions (which you must do since Fx is 1d based) on the server side

Reply:
I thought about this one. Since the Range can take in line number and char (zero indexing), 
![image](https://github.com/microsoft/Power-Fx/assets/98551644/004132fc-351d-4ca5-a139-f7ae999fee2c)
 I thought we could pass it in this format from the front end.

Reply:
I am not pointing to structuring the range. I am taking about the a position in an range being 1d index or 2d index (row and column vs flatten index)

An 1d index 5 points to 6th character in expression (0 based)
A 2d representation of 1d index 5 is line (row): 1 and column 5

Go from Range(Start: Position(line, row), End: Position(line, row)) to Range(Start: Index, End: Index). you can easily and very conveniently calculate 1d indices on the client side as opposed to server

Reply:
I see what you mean, in that case we can follow the same indexing like the one used in publishcontroltokens for control highlighting work before. 
I can keep it as unit[]   // Eg: [0,0,0,6] >> start line 0, start character 0, end line 0, end character 6.

Reply:
uint[] was done for response compression for semantic tokens. Here, you should introduce a class called 1DRange (better name please) and have two props inside it

class 1DRange {
  public int Start {get; set;}

  public int End {get; set;} 

}

and then use this type for your "range" param

on the UFB side of things, do model.getOffsetAt(2DRange) to get 1d start and 1d end